### PR TITLE
[#39] Pretty exceptions

### LINKED ISSUE #39: Implement pretty 'Show' instance for bi-exceptions

### PR DESCRIPTION
**Pull request by ChShersh**
Resolves #39

### CODE REVIEW

File: src/Toml/Bi/Code.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+import qualified Data.Text as Text
+
+-- | Type of exception for converting from 'Toml' to user custom data type.
+data DecodeException
+    = KeyNotFound Key  -- ^ No such key
+    | TableNotFound Key  -- ^ No such table
+    | TypeMismatch Key Text ValueType -- ^ Expected type vs actual type
+    | ParseError ParseException  -- ^ Exception during parsing
+    deriving (Eq, Show)  -- TODO: manual pretty show instances
```
Comment (vrom911) [inline src/Toml/Bi/Code.hs]: Isn't `show` is just `toString . prettyException` now?

File: src/Toml/Bi/Combinators.hs [hunk 2]
```diff
@@ -250,9 +213,13 @@ maybeP converter key = let bi = converter key in Bijection
...
     handleNotFound :: DecodeException -> Env (Maybe a)
-    handleNotFound (KeyNotFound _)   = pure Nothing
-    handleNotFound (TableNotFound _) = pure Nothing
-    handleNotFound e                 = throwError e
+    handleNotFound e@(KeyNotFound name)
+        | name == key = pure Nothing
+        | otherwise = throwError e
+    handleNotFound e@(TableNotFound name)
+        | name == key = pure Nothing
+        | otherwise = throwError e
```
Comment (vrom911) [inline src/Toml/Bi/Combinators.hs]: Probably we could some how get rid of code duplication.. But I can't see the better way for this, so let keep it as it is. At least it's readable!

File: src/Toml/Bi/Combinators.hs [hunk 3]
```diff
@@ -272,7 +239,7 @@ table bi key = Bijection input output
...
         a <$ modify (\(TOML vals tables) -> TOML vals (Prefix.insert key newToml tables))
 
-    rethrowNotFound :: DecodeException -> Env a
-    rethrowNotFound (KeyNotFound name)   = throwError $ IncompleteTable key name
-    rethrowNotFound (TableNotFound name) = throwError $ IncompleteTable key name
-    rethrowNotFound e                    = throwError e
+    addTableName :: DecodeException -> Env a
+    addTableName (KeyNotFound name)   = throwError $ KeyNotFound (key <> name)
+    addTableName (TableNotFound name) = throwError $ TableNotFound (key <> name)
+    addTableName e                    = throwError e
```
Comment (vrom911) [inline src/Toml/Bi/Combinators.hs]: Probably it's better to call it `tableNameHandler` or something like this just to see that it's for error handling?

Comment (ChShersh) [reply src/Toml/Bi/Combinators.hs]: Good catch! I also forgot about one error here...

File: src/Toml/Bi/Combinators.hs (hunk 2 above)
Comment (ChShersh) [reply src/Toml/Bi/Combinators.hs]: I think it's possible. Will try to rewrite it.

File: src/Toml/Bi/Code.hs (hunk 1 above)
Comment (ChShersh) [reply src/Toml/Bi/Code.hs]: Personally I like to keep default definition of `Show` just to be able to see structure of data type if needed. But I don't mind writing `Show` using `prettyException`. Though, it requires more code than just generating `Show` instance...

File: src/Toml/Bi/Combinators.hs [hunk 4]
```diff
@@ -213,13 +213,9 @@ maybeP converter key = let bi = converter key in Bijection
     }
   where
     handleNotFound :: DecodeException -> Env (Maybe a)
-    handleNotFound e@(KeyNotFound name)
-        | name == key = pure Nothing
+    handleNotFound e
+        | e `elem` [KeyNotFound key, TableNotFound key] = pure Nothing
```
Comment (vrom911) [inline src/Toml/Bi/Combinators.hs]: Wow! This is cool solution! 🍾

**Pull request merged**